app/controllers/reports_controller.rb

```diff
@@ -17,11 +17,34 @@ def update
     time_regs = TimeReg.for_report(client_ids_for_report, project_ids_for_report, user_ids_for_report, task_ids_for_report)
                        .where(date_worked: (@selected_start_date..@selected_end_date))
 
-    @structured_report_data = TimeRegsPresenter.new(time_regs).report_data(
-      title: nil,
-      keys: [ :client, :project, :task, :user ]
-    )
+    if @form_data.detailed_report
+
+      @structured_report_data = time_regs.group_by { |reg| reg[:date_worked] }.sort_by { |key| key  }
+
+      clients = Client.joins(:time_regs).where(time_regs: { id: time_regs }).distinct
+      projects = Project.joins(:time_regs).where(time_regs: { id: time_regs }).distinct
+      tasks = Task.joins(:time_regs).where(time_regs: { id: time_regs }).distinct
+      users = User.joins(:time_regs).where(time_regs: { id: time_regs }).distinct
+      total_billable_minutes = time_regs.joins(:project).where(project: { billable_project: true }).sum(:minutes)
+      total_minutes = time_regs.sum(:minutes)
+
+      @detailed_report_data = OpenStruct.new(
+        time_regs: time_regs,
+        total_billable_minutes: total_billable_minutes,
+        total_minutes: total_minutes,
+        clients: clients,
+        projects: projects,
+        tasks: tasks,
+        users: users,
+      )
```

Comment:
Great work here, @str1fe I agree with you I think some of the business logic could be moved to models and a bit of OOP to dry it up

Another point is that I've noticed multiple requests on each `check_box` tick. Here I'd propose to have it like a common filter view with one `update` button that submits it all. The reason is to reduce unnecessary network requests, we may opt to keep it this way but I think we'd need something like https://www.npmjs.com/package/debounce to delay function calls until a set time elapses after the last invocation and also a bit of caching where already fetched.

<img width="1920" alt="Screenshot 2024-03-21 at 16 38 50" src="https://github.com/rubynor/reap/assets/42216593/7af8466b-b4a3-4641-9f75-ffcce3b40e91">


PS: the current implementation doesn't hurt the performance or any server, I'm just leaving this here so that we can take care of it at some point

Reply:
or perhaps using service objects to reduce the size of the controller? here's an interesting read about it https://www.toptal.com/ruby-on-rails/rails-service-objects-tutorial